Deflake test stream translator web socket server errors

aojea opened the pull request:
/kind flake

```release-note
NONE
```

Fixes: #132506

First commit use subtests for better debugging

```
stress ./proxy.test -test.run TestStreamTranslator_BlockRedirects
5s: 1878 runs so far, 0 failures
10s: 3509 runs so far, 0 failures
15s: 4545 runs so far, 0 failures
20s: 5442 runs so far, 0 failures
25s: 6279 runs so far, 0 failures
```

Second commit checks asynchronous to avoid the race until the metric is updated


```
 stress ./proxy.test -test.run TestStreamTranslator_WebSocketServerErrors
tress ./proxy.test -test.run TestStreamTranslator_WebSocketServerErrors
5s: 5013 runs so far, 0 failures
10s: 9818 runs so far, 0 failures
15s: 14559 runs so far, 0 failures
20s: 19167 runs so far, 0 failures
25s: 23483 runs so far, 0 failures
30s: 27529 runs so far, 0 failures
35s: 30932 runs so far, 0 failures
40s: 34357 runs so far, 0 failures
45s: 37631 runs so far, 0 failures
50s: 40886 runs so far, 0 failures
55s: 43955 runs so far, 0 failures
1m0s: 46955 runs so far, 0 failures
1m5s: 50931 runs so far, 0 failures
1m10s: 55298 runs so far, 0 failures
```

/assign @liggitt

aojea committed:
TestStreamTranslator_BlockRedirects use subtests

Reset metrics per status code also allow to  log the particular status
code that flake, in case the problem comes from an specific one.

Change-Id: I29f67d4c5d76449350f45049f45ce8325a2d0ddd

aojea committed:
fix flake on TestStreamTranslator_WebSocketServerErrors

The metrics assertion race with the metric update, and since this
happens at the serverside, we use an active look to check the metrics
instead of expecting to be updated immidiatly.

Change-Id: I9a64b66301d5f4ac3df0c0a01de10602a20f89ea

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go:
```diff
@@ -755,90 +755,98 @@ func TestStreamTranslator_WebSocketServerErrors(t *testing.T) {
 `
-	if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...); err != nil {
-		t.Fatal(err)
+	if err := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 2*time.Second, true, func(ctx context.Context) (bool, error) {
```
we have to poll because the implementation is calling `metrics.IncStreamTranslatorRequest` _after_ writing the response to the client stream with `websocketStreams.writeStatus`, so the `errorChan <- exec.StreamWithContext` can possibly complete before the metrics get bumped?

would it make more sense to increment the metrics _first_ before writing the client status error response, rather than compensating with a poll in the test, e.g. 

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go
index 6593a3ed966..5ef8f018c7d 100644
--- a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go
@@ -24,6 +24,7 @@ import (
 	"time"
 
 	"github.com/mxk/go-flowrate/flowrate"
+
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
@@ -73,14 +74,14 @@ func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 	// Creating SPDY executor, ensuring redirects are not followed.
 	spdyRoundTripper, err := spdy.NewRoundTripperWithConfig(spdy.RoundTripperConfig{UpgradeTransport: h.Transport, PingPeriod: 5 * time.Second})
 	if err != nil {
-		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		return
 	}
 	spdyExecutor, err := remotecommand.NewSPDYExecutorRejectRedirects(spdyRoundTripper, spdyRoundTripper, "POST", h.Location)
 	if err != nil {
-		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		return
 	}
 
@@ -121,27 +122,27 @@ func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 	if err != nil {
 		//nolint:errcheck   // Ignore writeStatus returned error
 		if statusErr, ok := err.(*apierrors.StatusError); ok {
-			websocketStreams.writeStatus(statusErr)
 			// Increment status code returned within status error.
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(int(statusErr.Status().Code)))
+			websocketStreams.writeStatus(statusErr)
 		} else if exitErr, ok := err.(exec.CodeExitError); ok && exitErr.Exited() {
-			websocketStreams.writeStatus(codeExitToStatusError(exitErr))
 			// Returned an exit code from the container, so not an error in
 			// stream translator--add StatusOK to metrics.
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
+			websocketStreams.writeStatus(codeExitToStatusError(exitErr))
 		} else {
-			websocketStreams.writeStatus(apierrors.NewInternalError(err))
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+			websocketStreams.writeStatus(apierrors.NewInternalError(err))
 		}
 		return
 	}
 
+	metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
 	// Write the success status back to the WebSocket client.
 	//nolint:errcheck
 	websocketStreams.writeStatus(&apierrors.StatusError{ErrStatus: metav1.Status{
 		Status: metav1.StatusSuccess,
 	}})
-	metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
 }
 
 // translatorSizeQueue feeds the size events from the WebSocket
```

aojea replied:
ok, for the other tests it will work , but in this case the metrics always happen AFTER it returned, we need to keep it this way for this test, with the async loop

```
	websocketStreams, err := webSocketServerStreams(req, w, h.Options)
	if err != nil {
		// Client error increments bad request status code.
		metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusBadRequest))
		return
	}
```

liggitt replied:
ack

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go:
```diff
@@ -755,90 +755,98 @@ func TestStreamTranslator_WebSocketServerErrors(t *testing.T) {
-			if err == nil {
-				t.Fatalf("expected error, but received none")
+		t.Run(fmt.Sprintf("statusCode=%d", statusCode), func(t *testing.T) {
+			metrics.ResetForTest()
```
reset before *and* after the test?

```
metrics.ResetForTest()
t.Cleanup(metrics.ResetForTest)
```

aojea replied:
not really needed, the next subtest will reset them before start

aojea committed:
websocket streamtranslator increament metrics before writing status

Test that assert on metrics use the returned status to sync the test, if
the metric is updated AFTER the state is returned the test will flake
because the assertion on the metric will race.

Change-Id: I317708a22cb47256c37dac3cab0463a2f925ad6b

liggitt reviewed (commented):
/lgtm
/approve
/milestone v1.34

stabilizing test flakes for 1.34 release
